### PR TITLE
Spec for opcode `SHR`

### PR DESCRIPTION
**Pull request by silathdiir**

### CODE REVIEW

File: src/zkevm_specs/evm/execution/shr.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    a64s: Sequence[FQ],
+    b64s: Sequence[FQ],
+    a64s_lo: Sequence[FQ],
+    a64s_hi: Sequence[FQ],
+    shf_div64,
+    shf_mod64,
+    p_lo,
+    p_hi,
+):
+    shf_lt256 = instruction.is_zero(instruction.bytes_to_fq(shift.le_bytes[1:]))
```
Comment (icemelon) [inline src/zkevm_specs/evm/execution/shr.py:60]:
```suggestion
    shf_lt256 = instruction.is_zero(instruction.sum(shift.le_bytes[1:]))
```

File: src/zkevm_specs/evm/execution/shr.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    a64s_lo = [FQ(0)] * 4
+    a64s_hi = [FQ(0)] * 4
+    for idx in range(4):
+        a64s_lo[idx] = FQ(a64s[idx].n % p_lo.n)
+        a64s_hi[idx] = FQ(a64s[idx].n // p_lo.n)
+
+    b64s = [FQ(0)] * 4
+    b64s[3 - shf_div64.n] = a64s_hi[3]
+    for k in range(0, 3 - shf_div64.n):
+        b64s[k] = FQ(a64s_hi[k + shf_div64.n] + a64s_lo[k + shf_div64.n + 1] * p_hi.n)
```
Comment (icemelon) [inline src/zkevm_specs/evm/execution/shr.py:133]: To simplify this part a bit, you can first compute `bb = a.int_value >> shf0.n` and then break U256 value `bb` into 4 U64s `b64s`.

File: specs/opcode/1cSHR.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+Pop two EVM words `a` and `shift` from the stack, and push `b` to the stack, where `b` is computed as:
+
+1. If `shift >= 256`，then `b` is set to zero.
+2. If `shift < 256`，compute `b = a >> shift`.
+
+### Circuit behavior
+
+To prove the `SHR` opcode, we first construct a `ShrGadget` that proves `a >> shift = b` where `a, b, shift` are all 256-bit words.
+As usual, we use 32 cells to represent word `a` and `b`, where each cell holds a 8-bit value. Then split each word into four 64-bit limbs denoted by `a64s[idx]` and `b64s[idx]` where idx in `(0, 1, 2, 3)`.
+We put the lower `n` bits of a limb into the `lo` array, and put the higher `64 - n` bits into the `hi` array. During the SHR operation, the `lo` array will move to higher bits of the result, and the `hi` array will move to lower bits of the result.
```
Comment (icemelon) [inline specs/opcode/1cSHR.md:18]:
```suggestion
We put the lower `n` bits of a limb into the `lo` array, and put the higher `64 - n` bits into the `hi` array, where `n` is `shift % 64`. During the SHR operation, the `lo` array will move to higher bits of the result, and the `hi` array will move to lower bits of the result.
```

File: specs/opcode/1cSHR.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
++-------------------------------+-------------------------------+-----
+|             a64s[0]           |             a64s[1]           | ...
++------------+------------------+------------+------------------+-----
+| a64s_lo[0] |    a64s_hi[0]    | a64s_lo[1] |    a64s_hi[1]    | ...
++------------+------------------+------------+------------------+-----
+             |            b64s[0]            |            b64s[1]
+             +-------------------------------+------------------------
+```
+
+#### Variable Definition
```
Comment (icemelon) [inline specs/opcode/1cSHR.md:34]: remove this line

File: specs/opcode/1cSHR.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+a64s\_lo[idx] &= a64s[idx] ~\%~p\_lo \\
+a64s\_hi[idx] &= a64s[idx] ~/~ p\_lo \\
+\end{align}
+$$
+
+If $shift\ge 256$, `b64s` are all 0. Otherwise, `b64s` can be calculated by `a >> shf0` then split into four 64-bit limbs.
+
+#### Constraint Validation
+
+Then we could validate the below constraints:
```
Comment (icemelon) [inline specs/opcode/1cSHR.md:56]: Now putting things together, the constraints can be constructed as follows:

**Review by icemelon — COMMENTED** [review]
Otherwise lgtm

**Review by miha-stopar — APPROVED** [review]
LGTM! (some minor formatting issues mentioned in the comments)

File: specs/opcode/1cSHR.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+shf\_div64 &= shift // 64 \\
+shf\_mod64 &= shift \% 64 \\
+shf\_lt256 &= is\_zero(sum(shift[1:])) \\
+p\_lo &= 2^{shf\_mod64} \\
+p\_hi &= 2^{64 - shf\_mod64} \\
+a64s[idx] &= \sum_{i=0}^7 a[8\cdot idx + i] \cdot 256^i~~~~(idx=0,1,2,3) \\
+a64s\_lo[idx] &= a64s[idx] ~\%~p\_lo \\
+a64s\_hi[idx] &= a64s[idx] ~/~ p\_lo \\
+\end{align}
+$$
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply specs/opcode/1cSHR.md]: Convert this part as below:
```
shf0 = bytes_to_fq(shift.le_bytes[:1])
shf_div64 = shift // 64
shf_mod64 = shift % 64
shf_lt256 = is_zero(sum(shift[1:]))
p_lo = 1 << shf_mod64
p_hi = 1 << (64 - shf_mod64)
a64s = word_to_64s(a)
a64s_lo[idx] = a64s[idx] % p_lo
a64s_hi[idx] = a64s[idx] / p_lo
```

File: specs/opcode/1cSHR.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+(a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq0 +
+  a64s_hi[3] * shf_div64_eq1
+```
+
+* `b64s[3]` should be equal to:
+```
+a64s_hi[3] * shf_div64_eq0
+```
+
+4. `shift[0]` constraint
```
[the remark this answers is not in the document]
Comment (silathdiir) [reply specs/opcode/1cSHR.md]: Added a colon at the end of these numbered titles.

**Review by icemelon — APPROVED** [review]
LGTM. will wait for @miha-stopar to take another look

**Comment by miha-stopar** [comment]
> LGTM. will wait for @miha-stopar to take another look

Seems all good to me!

**Comment by icemelon** [comment]
I will merge this PR after taking a look at the circuit.

**Pull request merged**